backend/models/form.py

```diff
@@ -47,11 +48,27 @@ def validate_features(cls, value: list[str]) -> t.Optional[list[str]]:
         if any(v not in allowed_values for v in value):
             raise ValueError("Form features list contains one or more invalid values.")
 
-        if FormFeatures.COLLECT_EMAIL in value and FormFeatures.REQUIRES_LOGIN not in value:  # noqa
-            raise ValueError("COLLECT_EMAIL feature require REQUIRES_LOGIN feature.")
+        if FormFeatures.REQUIRES_LOGIN not in value:
+            if FormFeatures.COLLECT_EMAIL in value:
+                raise ValueError(
+                    "COLLECT_EMAIL feature require REQUIRES_LOGIN feature."
+                )
+
+            if FormFeatures.ASSIGN_ROLE in value:
+                raise ValueError("ASSIGN_ROLE feature require REQUIRES_LOGIN feature.")
 
         return value
 
+    @root_validator
+    def validate_role(cls, values: dict[str, t.Any]) -> t.Optional[dict[str, t.Any]]:
```

Comment:
The problem with having this as a validator is that: Validation is run on form submission, with admin being False, meaning that this non-public field will be not set, and cause a server error. You could probably figure out a solution to make it check only if the form is being added or something, but it would be simplest to just make the field public.